### PR TITLE
Feature/wellcost

### PR DESCRIPTION
**Pull request by bragesk**
WellCost feature.
UseWellCost = true (defaults to false), the parser will look for separate price for horizontal (XY) and vertical (Z) prices.
Defined by:
i.e.
"WellCostXY": 10000, (defaults to 0)
"WellCostZ": 5000 (defaults to 0)

If "SeparateHorizontalAndVertical": false, it will look for a general cost.

i.e.
"WellCost" = 10000 (defaults to 0)

### CODE REVIEW

File: FieldOpt/Model/model.h [hunk 1]
```diff
@@ -121,6 +125,22 @@ class Model : public Loggable
    private:
     Model *model_;
   };
+ public:
+  class economy{
```
Comment (einar90) [inline FieldOpt/Model/model.h]: * Class and struct names should have Capital First Letter
* IMO this should be a `struct`, not a `class`. A struct is typically used when all member attributes are public (directly accessible) (and when the object acts primarily as a container and doesn't to much else by itself)

File: FieldOpt/Model/model.h [hunk 2]
```diff
@@ -121,6 +125,22 @@ class Model : public Loggable
...
+    map<string, double> well_lengths;
+    double n_wells = 0;
+    double costXY = 0;
+    double costZ = 0;
+    double cost = 0;
+    bool separate = false;
+    bool useWellCost = false;
+    QList<Wells::Well *> wells_;
+  };
+  economy wellCost(Settings::Optimizer::Objective);
```
Comment (einar90) [inline FieldOpt/Model/model.h]: Public methods should have Capital First Letter (and yes, I know I haven't done a fantastic job keeping to this convention myself everywhere)

File: FieldOpt/Model/model.cpp [hunk 3]
```diff
@@ -105,6 +105,41 @@ void Model::verifyWells()
     }
 }
 
+Model::economy Model::wellCost(Settings::Optimizer::Objective objective) {
+  economy wellEconomy;
```
Comment (einar90) [inline FieldOpt/Model/model.cpp]: Attributes, both local (inside methods) and class members, should use `snake_case`. Member attributes should have a `trailing_underscore_`.

File: FieldOpt/Model/model.cpp [hunk 4]
```diff
@@ -105,6 +105,41 @@ void Model::verifyWells()
...
+  if (objective.useWellCost) {
+    wellEconomy.n_wells = wells_->size();
+    wellEconomy.cost = objective.wellCost;
+    wellEconomy.costXY = objective.wellCostXY;
+    wellEconomy.costZ = objective.wellCostZ;
+    wellEconomy.separate = objective.separatehorizontalandvertical;
+    wellEconomy.useWellCost = objective.useWellCost;
+    wellEconomy.wells_ = *wells_;
+    for (Wells::Well *well : *wells_) {
```
Comment (einar90) [inline FieldOpt/Model/model.cpp]: Just a tip: you can replace `Wells::Well *well` with `auto well`

File: FieldOpt/Model/model.cpp [hunk 5]
```diff
@@ -105,6 +105,41 @@ void Model::verifyWells()
...
+      auto well_spline_toe_y = variable[1]->y->value();
+      auto well_spline_toe_z = variable[1]->z->value();
+
+      wellEconomy.well_xy[well->name().toStdString()] = sqrt(
+          pow((well_spline_heel_x - well_spline_toe_x), 2) + pow((well_spline_heel_y - well_spline_toe_y), 2));
+      wellEconomy.well_z[well->name().toStdString()] = abs(well_spline_heel_z - well_spline_toe_z);
+      wellEconomy.well_lengths[well->name().toStdString()] = sqrt(
+          pow((well_spline_heel_x - well_spline_toe_x), 2) + pow((well_spline_heel_y - well_spline_toe_y), 2)
+              + pow((well_spline_heel_z - well_spline_toe_z), 2));
+
```
Comment (einar90) [inline FieldOpt/Model/model.cpp]: Unnecessary blank linebreaks from here to end of method; remove them.

File: FieldOpt/Optimization/objective/NPV.cpp [hunk 6]
```diff
@@ -104,13 +108,24 @@ double NPV::value() const {
         double prop_coeff = components_->at(i)->coefficient;
       }
     }
+    if (model_economics.useWellCost) {
+      for (Model::Wells::Well *well: model_economics.wells_) {
```
Comment (einar90) [inline FieldOpt/Optimization/objective/NPV.cpp]: Use `auto well`

File: FieldOpt/Optimization/objective/NPV.cpp [hunk 7]
```diff
@@ -104,13 +108,24 @@ double NPV::value() const {
...
+      }
+    }
     return value;
   }
   catch (...) {
     Printer::error("Failed to compute NPV. Returning 0.0");
     return 0.0;
   }
 
+
```
Comment (einar90) [inline FieldOpt/Optimization/objective/NPV.cpp]: Unnecessary whitespace

File: FieldOpt/Settings/optimizer.h [hunk 8]
```diff
@@ -82,6 +82,11 @@ class Optimizer
   struct Objective {
     ObjectiveType type; //!< The objective definition type (e.g. WeightedSum)
     bool use_penalty_function; //!< Whether or not to use penalty function (default: false).
+    bool useWellCost; //!<Whether to include a well cost associated with the length
```
Comment (einar90) [inline FieldOpt/Settings/optimizer.h]: `snake_case`

File: FieldOpt/Runner/runners/synchronous_mpi_runner.cpp [hunk 9]
```diff
@@ -215,7 +215,7 @@ void SynchronousMPIRunner::Execute() {
                 if (simulation_success) {
                     tag = MPIRunner::MsgTag::CASE_EVAL_SUCCESS;
                     printMessage("Setting objective function value.", 2);
-                    worker_->GetCurrentCase()->set_objective_function_value(objective_function_->value());
+                    worker_->GetCurrentCase()->set_objective_function_value(objective_function_->value(model_->wellCost(settings_->optimizer()->objective())));
```
Comment (einar90) [inline FieldOpt/Runner/runners/synchronous_mpi_runner.cpp]: I'm not happy with how value has to be called here. We should discuss better alternatives (nothing comes to mind right now, but I'll think about it)

File: FieldOpt/Optimization/objective/NPV.h [hunk 10]
```diff
@@ -64,6 +65,8 @@ class NPV : public Objective {
   QList<Component *> *components_; //!< List of gamma, k pairs.
   Simulation::Results::Results *results_;  //!< Object providing access to simulator results.
   Settings::Optimizer *settings_;
+  Model::Model::Economy *well_economy_;
+  Model::Model *model_;
```
Comment (einar90) [inline FieldOpt/Optimization/objective/NPV.h]: Is it necessary to keep the model? Isn't WellEconomy enough?

File: FieldOpt/Runner/runners/oneoff_runner.cpp [hunk 11]
```diff
@@ -22,6 +22,7 @@ void OneOffRunner::Execute()
     applyWellPositionFromArguments();
     EvaluateBaseModel();
     // Write objective function value to file
+    model_->wellCost(settings_->optimizer()->objective());
     Utilities::FileHandling::WriteLineToFile(QString("%1").arg(objective_function_->value()),
```
Comment (einar90) [inline FieldOpt/Runner/runners/oneoff_runner.cpp]: I know this isn't your code Brage, but when on earth did this slip in? What's the point? It's already written to the case log...

File: FieldOpt/Settings/optimizer.h [hunk 12]
```diff
@@ -82,6 +82,11 @@ class Optimizer
   struct Objective {
     ObjectiveType type; //!< The objective definition type (e.g. WeightedSum)
     bool use_penalty_function; //!< Whether or not to use penalty function (default: false).
+    bool use_well_cost; //!<Whether to include a well cost associated with the length
+    bool separatehorizontalandvertical;
```
Comment (einar90) [inline FieldOpt/Settings/optimizer.h]: You should add some more `//!< Doxygen comments` here, e.g. is the cost pr. meter?

File: FieldOpt/Model/model.h [hunk 13]
```diff
@@ -121,6 +121,22 @@ class Model : public Loggable
    private:
     Model *model_;
   };
+ public:
+  struct Economy{
```
Comment (einar90) [inline FieldOpt/Model/model.h]: Add a
```
/*! \brief Economy struct does blah blah
 * 
 * To be specific, the cost represents blah, and has unit foo pr. bar.
 */
```
Doxygen comment.

File: FieldOpt/Runner/runners/oneoff_runner.cpp (hunk 11 above)
Comment (bragesk) [reply FieldOpt/Runner/runners/oneoff_runner.cpp]: Line 25 or 26? I have no idea, I no touchy line 26

File: FieldOpt/Optimization/objective/NPV.h (hunk 10 above)
Comment (bragesk) [reply FieldOpt/Optimization/objective/NPV.h]: Yupp

File: FieldOpt/Model/model.cpp [hunk 14]
```diff
@@ -105,6 +105,36 @@ void Model::verifyWells()
...
+  if (objective.use_well_cost) {
+    well_economy.n_wells = wells_->size();
+    well_economy.cost = objective.wellCost;
+    well_economy.costXY = objective.wellCostXY;
+    well_economy.costZ = objective.wellCostZ;
+    well_economy.separate = objective.separatehorizontalandvertical;
+    well_economy.use_well_cost = objective.use_well_cost;
+    well_economy.wells_ = *wells_;
+    for (Wells::Well *well : *wells_) {
+      auto variable = well->trajectory()->GetWellSpline()->GetSplinePoints();
```
[the remark this answers is not in the document]
Comment (bragesk) [reply FieldOpt/Model/model.cpp]: Although I see your point about well_xy, the use case you suggest here is not what it was designed for. Rather, it is specified in case you have different costs associated with drilling horizontal and vertical.
